tracks/javascript/exercises/hamming/mentoring.md

```diff
@@ -0,0 +1,62 @@
+This exercise focuses on:
+- control flow loops: how to _iterate_ through two collections and how to _count_ based on a condition
+
+## Reasonable solutions
+```javascript
+export function compute(leftStrand, rightStrand) {
+  const { length } = leftStrand;
+  if (length !== rightStrand.length) {
+    throw Error('left and right strands must be of equal length');
+  }
+
+  let distance = 0;
+  for (let i = 0; i < length; i++) {
+    distance += leftStrand.charCodeAt(i) !== rightStrand.charCodeAt(i);
+  }
+  return distance;
+}
+```
+Variations include `String#charAt(i)` and `String#[i]` (similar speed)
+
+Using `continue` (similar speed):
+```javascript
+if (leftStrand.charCodeAt(i) === rightStrand.charCodeAt(i)) {
+  continue;
+}
+distance += 1;
+```
+
+Explicit addition with ternary (similar speed):
+```javascript
+distance += leftStrand[i] === rightStrand[i] ? 0 : 1
+```
+Variations include re-assignment using `? distance : distance + 1`
+
+### Using streams
+Vanilla JavaScript doesn't have a `Array#zip(array)` or `Enumerable#count(expression)` function, yet, but it can still be attempted by using `split` with `reduce` or `replace`. These solutions work, but are a factor 5-10 slower. You might want to challenge a student to use a for loop. **Note** that performance is heavily impacted by the JavaScript engine and things like bable/bubble implementation/version.
```

Comment:
If those functions do not exist in vanilla JavaScript, were do they exist then?

Reply:
I think it is easier for the reader if you either mention those (or even link to them) in a side note or don't mention it at all.

Also raw speed is often not the (first) thing to optimize for. Until a bottleneck has been identified one should write code that is easily understood by the team. (Of course FP style code is not considered readable by most people ;) )

Reply:
(p.s. the ruby notes I use and mentoring I do is exactly what you suggest: use stdlib functional style and try to remove all bookkeeping variables.

I still add the streams solution to these notes because they are valid and readable and I prefer them :p. "Might want to challenge" for me is literally that.)

Reply:
> Pulling in libraries to fulfill tasks you can do with a simple for loop is not something we want to encourage. Dependency hell is a common problem in JavaScript.

Yupp, I understand. But you assume that every mentor who shall work with your text knows about the modules/classes/methods/whatever. This might not be true at all, therefore you should provide those explanation for the other mentors.

> Optimising for raw speed on the other hand ís what has been done (over the past years) in most mentor sessions on the JavaScript track.

Mentoring is only available for ~6 months now.

And if your current mentors are all arguing for speed, then you should propably encourage a shift of mind.

Of course I can understand, that especially for vanilla javascript, the most performant reason is also one of the most performant, but that is just a lucky coincident.

> Personally I think intuitively knowing that using a simple for loops is almost always faster is a good thing and should be taught rather sooner than later.

For most JS developers I had contact with so far I think that also a straight `for`ward loop (pun intended) is also the most readable way considering the more FP style approaches.

So I do not see anything bad in a `for` loop, but I don't like the priority that is given to raw execution speed in the mentor notes.

Reply:
I can definitely link to some of the fp style extensions. That is a good idea.

Mentoring has only been available for that long, but before that there was community commentary and as you can see I've been on here for four years. For some reason speed is always a thing. Then again, outside of exercism, premature optimisation with JavaScript had always been a thing.

When I'm mentoring I do approve readable answers over for loops -- I dont agree with you that _mentioning_ other solutions are slower is something we shouldn't have in the mentor notes. 

The issue with JavaScript is that on low end devices, a 2x or in this case up to 10x can actually be so significant it is visible :cry:. Of course not what these simple exercises (unless they make a O(2**n) implementation))…

Reply:
Funny thing though, myself I prefer the functional and streams approach almost always over a for loop when it comes to readability :flushed: